services/CurrentTime/source/CurrentTimeService.cpp

```diff
@@ -0,0 +1,236 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2020 ARM Limited
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "CurrentTimeService.h"
+
+#define CURRENT_TIME_CHAR_VALUE_SIZE 10
+#define DATA_FIELD_IGNORED 0x80
+
+using namespace std::literals::chrono_literals;
+
+CurrentTimeService::CurrentTimeService(BLE &ble, events::EventQueue &event_queue) :
+    _ble(ble),
+    _event_queue(event_queue),
+    _current_time_char(GattCharacteristic::UUID_CURRENT_TIME_CHAR, &_current_time)
+{
+}
+
+ble_error_t CurrentTimeService::init()
+{
+    GattCharacteristic *charTable[] = {&_current_time_char};
+    GattService currentTimeService(GattService::UUID_CURRENT_TIME_SERVICE, charTable, 1);
+
+    _current_time_char.setReadAuthorizationCallback (this, &CurrentTimeService::onCurrentTimeRead);
+    _current_time_char.setWriteAuthorizationCallback(this, &CurrentTimeService::onCurrentTimeWritten);
+
+    ble_error_t bleError = _ble.gattServer().addService(currentTimeService);
+
+    start_periodic_time_update();
+
+    MBED_STATIC_ASSERT(sizeof(_current_time) == CURRENT_TIME_CHAR_VALUE_SIZE, "Current time characteristic value size = 10");
+
+    return bleError;
+}
+
+void CurrentTimeService::set_event_handler(EventHandler *handler) {
+    _current_time_handler = handler;
+}
+
+time_t CurrentTimeService::get_time() const
+{
+    time_t epoch_time = time(nullptr);
+
+    return epoch_time + _time_offset;
+}
+
+void CurrentTimeService::set_time(time_t host_time, uint8_t adjust_reason)
+{
+    time_t epoch_time = time(nullptr);
+
+    time_t old_time_offset = _time_offset;
+    _time_offset = host_time - epoch_time;
+
+    update_current_time_value(adjust_reason, old_time_offset - _time_offset);
+}
+
+void CurrentTimeService::update_current_time_value(const uint8_t adjust_reason, time_t time_offset_difference) {
```

Comment:
Could you change `adjust_reason` into `adjust_reasons` ? This byte may contain several reasons. Please apply the same change where relevant.

Reply:
I guess but the field is not given a plural name in the spec. The value can signify a compound of reasons without having an 's' at the end.